### PR TITLE
cashapp Pay module/component

### PR DESCRIPTION
**Pull request by erenbesel**
<new>

* Payment Method: CashApp Pay. `paymentMethod.type`: **cashapp**

</new>

### CODE REVIEW

File: Adyen.docc/API Reference.md [hunk 1]
```diff
@@ -61,6 +61,7 @@ The Adyen DropIn/Components SDK API Reference.
 - ``StoredCardPaymentMethod``
 - ``StoredBCMCPaymentMethod``
 - ``StoredBLIKPaymentMethod``
+- ``StoredACHDirectDebitPaymentMethod``
```
Comment (descorp) [inline Adyen.docc/API Reference.md:64]: Is this related to `CashApp` ?

File: Adyen.docc/Installation.md [hunk 2]
```diff
@@ -63,13 +65,14 @@ You can add all modules or select individual modules to add to your integration.
...
 
-* `AdyenDropIn`: all modules except `AdyenWeChatPay`, `AdyenSwiftUI` and `AdyenSession`.
+* `AdyenDropIn`: all modules except `AdyenWeChatPay` `AdyenCashAppPay`, `AdyenSwiftUI` and `AdyenSession`.
 * `AdyenSession`: handler for the simplified checkout flow.
 * `AdyenCard`: the card components.
 * `AdyenComponents`: all other payment components except WeChat Pay.
 * `AdyenActions`:  action components.
 * `AdyenEncryption`: encryption.
 * `AdyenWeChatPay`: WeChat Pay component.
+* `AdyenCashAppPay`: Cash App Pay component.
```
Comment (descorp) [inline Adyen.docc/Installation.md:75]:
```suggestion
* `AdyenDropIn`: a pre-built UI solution for showing all payment methods as a list and accepting them in one flow.
* `AdyenSession`: fastest integration, requires you to make one API request to [/sessions](https://docs.adyen.com/api-explorer/Checkout/latest/post/sessions) endpoint.
* `AdyenComponents`: a collection of components for collecting shopper's payment data.
* `AdyenCard`: a collection of components for collecting card-based shopper's payment data.
* `AdyenActions`:  a collection of components for additional action handling.
* `AdyenEncryption`: a client side encryption helpers.
* `AdyenWeChatPay`: WeChat Pay component.
* `AdyenCashAppPay`: Cash App Pay component.
```

File: Adyen/Core/Payment Methods/CashAppPaymentMethod.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import Foundation
+
+/// Cash App Pay payment method
+public struct CashAppPayPaymentMethod: PaymentMethod {
+   
+    public let type: PaymentMethodType
+    
+    public let name: String
+    
+    public let clientId: String
```
Comment (descorp) [inline Adyen/Core/Payment Methods/CashAppPaymentMethod.swift:17]: This one need comment.

File: Adyen/Core/Payment Methods/CashAppPaymentMethod.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    
+    public let scopeId: String
```
Comment (descorp) [inline Adyen/Core/Payment Methods/CashAppPaymentMethod.swift:20]: Need comment.

File: Adyen/Model/CashAppPayConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Configuration for Cash App Pay Component.
+public struct CashAppPayConfiguration {
+
+    /// The URL for Cash App to call in order to redirect back to your application.
+    public let redirectURL: URL
```
Comment (descorp) [inline Adyen/Model/CashAppPayConfiguration.swift]: How is it working with session's returnUrl ?

---

Maybe we can consider feeding it to `AdyenContext` 🤔

File: Adyen/Model/StorePaymentMethodFieldAware.swift [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 //
 // This file is open source and available under the MIT license. See the LICENSE file for more info.
 //
 
 import Foundation
 
 @_spi(AdyenInternal)
-/// Describes the interface to have a store payment method field.
+/// Describes the interface to display a field to store the payment method.
 public protocol StorePaymentMethodFieldAware: AdyenSessionAware {
```
Comment (descorp) [inline Adyen/Model/StorePaymentMethodFieldAware.swift:11]: In next PR, we can apply this protocol to other components, right?

File: Adyen/UI/Form/FormItemViewBuilder.swift [hunk 7]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2022 Adyen N.V.
+// Copyright (c) 2023 Adyen N.V.
```
Comment (descorp) [inline Adyen/UI/Form/FormItemViewBuilder.swift:2]: This seems unnecessary 😅

File: Adyen/Core/Components/Installment/InstallmentOptions.swift [hunk 8]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2022 Adyen N.V.
+// Copyright (c) 2023 Adyen N.V.
```
Comment (descorp) [inline Adyen/Core/Components/Installment/InstallmentOptions.swift:2]: This seems unnecessary 😅

File: AdyenCard/Components/Card/CardComponent.swift [hunk 9]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2022 Adyen N.V.
+// Copyright (c) 2023 Adyen N.V.
```
Comment (descorp) [inline AdyenCard/Components/Card/CardComponent.swift:2]: This seems unnecessary 😅

File: AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md
Comment (descorp) [inline AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md:1]: I admire the effort! This is supper handy to have 🤩

File: AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
+# ``AdyenCashAppPay``
+
+This module is used for supporting the Cash App Pay payment method via the `CashAppPayComponent`. In order to support Cash App Pay, make sure to include this module in your application.
+You will also need a Client ID from Cash App, configured in your customer area. 
+
+## Overview
+
+The flow for Cash App Pay starts with the `CashAppPayComponent`. After pressing the button on the component, your app will redirect to the Cash App mobile app so the shopper can autherize the payment, after which the Cash App Pay SDK will return back to your app via the `redirectURL` you need to provide to the component.
```
Comment (descorp) [inline AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md]:
```suggestion
The flow for Cash App Pay starts with the `CashAppPayComponent`. After pressing the button on the component, your app will redirect to the Cash App mobile app so the shopper can authorise the payment, after which the Cash App Pay SDK will return back to your app via the `redirectURL` you need to provide to the component. 
```

File: AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+Present your component's view (modally, in a navigation stack, etc).
+```
+navigationController.present(component.viewController)
+```
+
+## Recurring Payment
+
+`CashAppPayComponent` supports recurring payments via its `configuration` property in `CashAppPayConfiguration` class. 
+If you are using the default integration with the `/sessions` call, you will need to set `recurringProcessingModel` and `storePaymentMethodMode` properties (v70+) ([API Explorer](https://docs.adyen.com/api-explorer/Checkout/70/post/sessions#request-storePaymentMethodMode))
+Showing the switch for storing the payment will be based on the response of the `/sessions` call. The switch will only be visible if you specified `AskForConsent` for `storePaymentMethodMode`. If you specify `Enabled` or `Disabled`, the switch will then be hidden. In which case, specific to this component, you will need to specifiy another flag in the `configuration` to store the payment method since Cash App SDK needs to know explicitly whether to create a recurring request:
```
Comment (descorp) [inline AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md]:
```suggestion
Showing the switch for storing the payment will be based on the response of the `/sessions` call. The switch will only be visible if you specified `AskForConsent` for `storePaymentMethodMode`. If you specify `Enabled` or `Disabled`, the switch will then be hidden. In which case, specific to this component, you will need to specify another flag in the `configuration` to store the payment method since Cash App SDK needs to know explicitly whether to create a recurring request:
```

File: AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+```
+- storePaymentMethod: To store the payment method. This property is ignored if `showsStorePaymentMethodField` is `true`.
+```
+
+If you are using the Advanced Flow (`/paymentMethods` endpoint instead of `/sessions`), then you can modify the `showsStorePaymentMethodField` yourself and if set to `false`, you will need to set `storePaymentMethod` as well for storing the payment method.
+
+After the above configurations, on the default flow, `AdyenSession` will make the required `/payments` call to autherize/store the payment.
```
Comment (descorp) [inline AdyenCashAppPay/AdyenCashAppPay.docc/AdyenCashAppPay.md]:
```suggestion
After the above configurations, on the default flow, `AdyenSession` will make the required `/payments` call to authorise/store the payment.
```

File: AdyenCashAppPay/CashAppPayComponent.swift [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        let cashAppPayKit = CashAppPay(clientID: cashAppPayPaymentMethod.clientId, endpoint: cashAppPayEnvironment)
+        cashAppPayKit.addObserver(self)
+        return cashAppPayKit
+    }()
+
+    private var cashAppPayEnvironment: CashAppPay.Endpoint {
+        guard let environment = context.apiContext.environment as? Environment else {
+            return .sandbox
+        }
+        return environment.isLive ? .production : .sandbox
```
Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift:68]: Just a thought:

with the same result, we can just check if `environment == .test` 🤔

File: AdyenCashAppPay/CashAppPayComponent.swift [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+
+    internal lazy var cashAppPayButton: FormButtonItem = {
+        let item = FormButtonItem(style: configuration.style.mainButtonItem)
+        item.identifier = ViewIdentifierBuilder.build(scopeInstance: self, postfix: ViewIdentifier.cashAppButtonItem)
+        item.title = localizedString(.cashAppPayTitle, configuration.localizationParameters)
+        item.buttonSelectionHandler = { [weak self] in
+            self?.didSelectSubmitButton()
+        }
+        return item
+    }()
```
Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift]: So we are not going to use CashApps own button here?

File: AdyenCashAppPay/CashAppPayComponent.swift [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+        cashAppPay.createCustomerRequest(params: CreateCustomerRequestParams(actions: actions,
+                                                                             redirectURL: configuration.redirectURL,
+                                                                             referenceID: configuration.referenceId,
+                                                                             metadata: nil))
+    }
+
+    private func createPaymentActions() -> [PaymentAction] {
+        var actions = [PaymentAction]()
+        if let amount = payment?.amount, amount.value > 0 {
+            let moneyAmount = Money(amount: UInt(amount.value), currency: .USD)
```
Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift:151]: Is `.USD` - the only currency available?

Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift:151]: Our `amount` is in minor units. is this ok for CashApp?

File: AdyenCashAppPay/CashAppPayComponent.swift [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+
+    public func stopLoading() {
+        cashAppPayButton.showsActivityIndicator = false
+        formViewController.view.isUserInteractionEnabled = true
+    }
+
+    private func startCashAppPayFlow() {
+        let actions = createPaymentActions()
+        guard actions.isEmpty == false else {
+            AdyenAssertion.assertionFailure(message: "At least one paymentAction is required to create a CashAppPay rquest")
```
Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift]: How is it working: if merchant provide no `payment`, but allows stored payments, then if shopper toggles "yes" - it will work. Otherwise it will do nothing (or crash the app in DEBUG)? 

Maybe better to do `didFail` instead? 🤔

Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift]: This error is not very descriptive:
Is it really clear to merchant what is "paymentAction" ?
How can one know that "oneTimePayment" was not created, because `payment` was not set to AdyenContext or that amount is zero?

File: AdyenCashAppPay/CashAppPayComponent.swift [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+        if let amount = payment?.amount, amount.value > 0 {
+            let moneyAmount = Money(amount: UInt(amount.value), currency: .USD)
+            let oneTimeAction = PaymentAction.oneTimePayment(scopeID: cashAppPayPaymentMethod.scopeId,
+                                                             money: moneyAmount)
+            actions.append(oneTimeAction)
+        }
+    
+        if storePayment {
+            let onFileAction = PaymentAction.onFilePayment(scopeID: cashAppPayPaymentMethod.scopeId,
+                                                           accountReferenceID: nil)
```
Comment (descorp) [inline AdyenCashAppPay/CashAppPayComponent.swift:159]: Here should be `referenceId` ?

File: AdyenDropIn/DropInComponent.swift [hunk 18]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2022 Adyen N.V.
+// Copyright (c) 2023 Adyen N.V.
```
Comment (descorp) [inline AdyenDropIn/DropInComponent.swift:2]: Not necessary

File: AdyenCashAppPay/CashAppPayDetails.swift [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    /// Grant id for recurring payments.
+    public let onFileGrantId: String?
+    
+    /// Unique identifier for this customer issued by Cash App.
+    public let customerId: String?
+    
+    /// Public identifier for the customer on Cash App.
+    public let cashtag: String?
+    
+    /// Creates and returns a Cash App Pay details instance.
```
Comment (descorp) [inline AdyenCashAppPay/CashAppPayDetails.swift]:
```suggestion
    /// Initializes an instance of `CashAppPayDetails`.
```

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift [hunk 20]
```diff
@@ -14,6 +14,9 @@
 #if canImport(AdyenActions)
     @_spi(AdyenInternal) import AdyenActions
 #endif
+#if canImport(AdyenCashAppPay)
+    @_spi(AdyenInternal) import AdyenCashAppPay
```
Comment (descorp) [inline AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift]: `@_spi(AdyenInternal)` here is not needed - we don't have any "hidden" methods or classes inside, right?

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift [hunk 21]
```diff
@@ -156,6 +159,28 @@ extension ComponentManager: PaymentComponentBuilder {
                      context: context,
                      configuration: .init(style: configuration.style.formComponent))
     }
+    
+    internal func build(paymentMethod: CashAppPayPaymentMethod) -> PaymentComponent? {
+        #if canImport(PayKit)
+            guard var cashAppPayConfig = configuration.cashAppPay else {
+                AdyenAssertion.assertionFailure(
+                    message: "Cash App Pay configuration instance must not be nil in order to use CashAppPayComponent")
```
Comment (descorp) [inline AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift:167]: Maybe better do 
```suggestion
                adyenPrint("Cash App Pay configuration instance must not be nil in order to use CashAppPayComponent")
```

File: Demo/Common/IntegrationExample.swift [hunk 22]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2022 Adyen N.V.
+// Copyright (c) 2023 Adyen N.V.
```
Comment (descorp) [inline Demo/Common/IntegrationExample.swift:2]: Not necessary

File: Demo/UIKit/AppDelegate.swift [hunk 23]
```diff
@@ -6,6 +6,7 @@
 
 import Adyen
 import AdyenActions
+import PayKit
```
Comment (descorp) [inline Demo/UIKit/AppDelegate.swift]: Maybe add `canImport` ? In case we are not using this for CashApp testing

File: Demo/UIKit/AppDelegate.swift [hunk 24]
```diff
@@ -34,6 +35,12 @@ internal final class AppDelegate: UIResponder, UIApplicationDelegate {
     internal func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
         RedirectComponent.applicationDidOpen(from: url)
```
Comment (descorp) [inline Demo/UIKit/AppDelegate.swift:38]: Just checking: This callback should't trigger anything on `RedirectComponent` ?

File: README.md [hunk 25]
```diff
@@ -362,5 +362,6 @@ This repository is open source and available under the MIT license. For more inf
 [reference.AtomeComponent]:  https://adyen.github.io/adyen-ios/5.0.0/documentation/adyen/atomecomponent
 [reference.UPIComponent]:  https://adyen.github.io/adyen-ios/5.0.0/documentation/adyen/upiComponent
 [reference.QRCodeActionComponent]:  https://adyen.github.io/adyen-ios/5.0.0/documentation/adyen/qrCodeActionComponent
+[reference.CashAppPayComponent]:  https://adyen.github.io/adyen-ios/5.0.0/documentation/adyen/cashAppPayComponent
```
Comment (descorp) [inline README.md:366]: This line doesn't do anything on it's own.

File: Tests/Components Tests/Cash App Pay/CashAppPayComponentTests.swift
Comment (descorp) [inline Tests/Components Tests/Cash App Pay/CashAppPayComponentTests.swift:1]: Would be nice to also add tests for DriopIn - that ComponentManager can create an instance

File: Adyen.docc/Installation.md (hunk 2 above)
Comment (erenbesel) [reply Adyen.docc/Installation.md:75]: That looks like it could be its own PR (the one where we include Session in the default PR)

File: Adyen.docc/API Reference.md (hunk 1 above)
Comment (erenbesel) [reply Adyen.docc/API Reference.md:64]: No just I realized it was not there, so I added it.

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 13 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift:68]: What about `local`, '`beta` etc?

File: AdyenCard/Components/Card/CardComponent.swift (hunk 9 above)
Comment (erenbesel) [reply AdyenCard/Components/Card/CardComponent.swift:2]: Not sure why they are changed, maybe I added a space somewhere and it reformatted them? But no reason to change them back right :)?

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 15 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift:151]: yes

Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift:151]: yes

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 16 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift]: It will go to Cash App for authorizing onFile (recurring) payment. 
As I mentioned, recurring work is not finished so we can't test the flow fully yet.

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 17 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift:159]: No referenceId and accountReferenceId are different things on cashapp

File: Demo/UIKit/AppDelegate.swift (hunk 24 above)
Comment (erenbesel) [reply Demo/UIKit/AppDelegate.swift:38]: No it does not 👍

File: Adyen.docc/Installation.md (hunk 2 above)
Comment (descorp) [reply Adyen.docc/Installation.md:75]: Fair enought!

File: AdyenCard/Components/Card/CardComponent.swift (hunk 9 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardComponent.swift:2]: well, it is a minor thing. I'll leave it up to you 🙂

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 13 above)
Comment (descorp) [reply AdyenCashAppPay/CashAppPayComponent.swift:68]: ok, fair point

File: Adyen/Model/CashAppPayConfiguration.swift (hunk 5 above)
Comment (erenbesel) [reply Adyen/Model/CashAppPayConfiguration.swift]: Hmm, curious. Let's discuss this in person

File: README.md (hunk 25 above)
Comment (erenbesel) [reply README.md:366]: Yeah doc isn't created before release, let's discuss this too.

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 16 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift]: Because they may only be going for pre auth with no amount and only set up recurring. In that case only onFile action would be needed and configuration needs to be set.
But checking those error cases in more details seemed redundant. It's only 2 element array and not very hard to see which one can be the cause. 

Open to discuss this one as well :)

File: AdyenCashAppPay/CashAppPayDetails.swift (hunk 19 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayDetails.swift]: Updated 2540dc919b172618b7ca124408e7507dd2d393d5

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift (hunk 20 above)
Comment (erenbesel) [reply AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift]: updated 2540dc919b172618b7ca124408e7507dd2d393d5

File: Demo/UIKit/AppDelegate.swift (hunk 23 above)
Comment (erenbesel) [reply Demo/UIKit/AppDelegate.swift]: updated 2540dc919b172618b7ca124408e7507dd2d393d5

File: Tests/Components Tests/Cash App Pay/CashAppPayComponentTests.swift
Comment (erenbesel) [reply Tests/Components Tests/Cash App Pay/CashAppPayComponentTests.swift:1]: nice one, added 2540dc919b172618b7ca124408e7507dd2d393d5

File: Adyen/Core/Payment Methods/CashAppPaymentMethod.swift (hunk 3 above)
Comment (erenbesel) [reply Adyen/Core/Payment Methods/CashAppPaymentMethod.swift:17]: 2540dc919b172618b7ca124408e7507dd2d393d5

File: Adyen/Core/Payment Methods/CashAppPaymentMethod.swift (hunk 4 above)
Comment (erenbesel) [reply Adyen/Core/Payment Methods/CashAppPaymentMethod.swift:20]: 2540dc919b172618b7ca124408e7507dd2d393d5

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 16 above)
Comment (descorp) [reply AdyenCashAppPay/CashAppPayComponent.swift]: How about making this checks in `throwing init` like we do for ApplePay?
This will prevent showing it on DropIn as well

File: Demo/UIKit/AppDelegate.swift (hunk 24 above)
Comment (descorp) [reply Demo/UIKit/AppDelegate.swift:38]: Now that I am thinking: `NotificationCenter.default.post` should also be wrapped 🤔

File: README.md (hunk 25 above)
Comment (descorp) [reply README.md:366]: No, I mean the Markdown `[some_name]: https://some.url.com` is just a reference. 
It must be used somewhere, like [here](https://github.com/Adyen/adyen-ios/blob/5bb1f9fd5b60c2006ee5b12acd06acbce479ac12/README.md?plain=1#L269)

File: AdyenActions/UI/View Controllers/Await/AwaitComponentViewModel.swift [hunk 26]
```diff
@@ -36,16 +36,21 @@ internal struct AwaitComponentViewModel {
     /// - Parameter localizationParameters: The localization parameters to control some aspects of how strings are localized
     internal static func viewModel(with paymentMethodType: AwaitPaymentMethod,
                                    localizationParameters: LocalizationParameters? = nil) -> AwaitComponentViewModel {
-        let localizationKey = LocalizationKey(key: "adyen.\(paymentMethodType.rawValue).confirmPayment")
+        
         switch paymentMethodType {
-        case .mbway, .blik:
+        case .blik:
```
Comment (neelSharma12) [inline AdyenActions/UI/View Controllers/Await/AwaitComponentViewModel.swift:41]: Nice and thank you! :)

File: AdyenCashAppPay/AdyenCashAppPay.h [hunk 27]
```diff
@@ -0,0 +1,19 @@
+//
+//  AdyenCashAppPay.h
+//  AdyenCashAppPay
+//
+//  Created by Eren Besel on 4/11/23.
+//  Copyright © 2023 Adyen. All rights reserved.
```
Comment (neelSharma12) [inline AdyenCashAppPay/AdyenCashAppPay.h:6]: This copyright header is different from our usual header.

File: README.md (hunk 25 above)
Comment (neelSharma12) [reply README.md:366]: I think we need to add this in `Available Components` as well. I did for UPI and QRCode.

File: Tests/Components Tests/Cash App Pay/CashAppPayComponentTests.swift [hunk 28]
```diff
@@ -0,0 +1,265 @@
+//
+//  CashAppPayComponentTests.swift
+//  AdyenUIKitTests
+//
+//  Created by Eren Besel on 4/17/23.
+//  Copyright © 2023 Adyen. All rights reserved.
```
Comment (neelSharma12) [inline Tests/Components Tests/Cash App Pay/CashAppPayComponentTests.swift]: This is not our usual copyright header. To avoid this should we make our own .plist file with our usual header and keep in the Xcode shared folder.

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift [hunk 29]
```diff
@@ -156,6 +159,28 @@ extension ComponentManager: PaymentComponentBuilder {
                      context: context,
                      configuration: .init(style: configuration.style.formComponent))
     }
+    
+    internal func build(paymentMethod: CashAppPayPaymentMethod) -> PaymentComponent? {
+        #if canImport(PayKit)
+            guard var cashAppPayConfig = configuration.cashAppPay else {
+                AdyenAssertion.assertionFailure(
```
Comment (descorp) [inline AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift:172]: Maybe `adyenPrint("Cash App Pay configuration instance must not be nil in order to use CashAppPayComponent")` ?
In case merchant is integrating with PayKit not via Adyen. In this case it should never be in `\paymentMethods`, hence never called, but still 🤔

File: Demo/Common/IntegrationExampleDropIn.swift [hunk 30]
```diff
@@ -52,6 +52,8 @@ extension IntegrationExample {
         configuration.actionComponent.threeDS.requestorAppURL = URL(string: ConfigurationConstants.returnUrl)
         configuration.card.billingAddress.mode = .postalCode
         configuration.paymentMethodsList.allowDisablingStoredPaymentMethods = true
+        configuration.cashAppPay = CashAppPayConfiguration(redirectURL: URL(string: "ui-host://callback")!,
```
Comment (descorp) [inline Demo/Common/IntegrationExampleDropIn.swift]: Maybe we should have new constant in `ConfigurationConstants` ?

File: Adyen/Core/Payment Methods/CashAppPaymentMethod.swift [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    
+    private enum ConfigurationCodingKeys: String, CodingKey {
+        case clientId
+        case scopeId
+    }
+}
+
+/// A stored Cash App Pay account.
+public struct StoredCashAppPayPaymentMethod: StoredPaymentMethod {
```
Comment (descorp) [inline Adyen/Core/Payment Methods/CashAppPaymentMethod.swift]: Maybe move to a dedicated file?

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift [hunk 32]
```diff
@@ -39,6 +42,12 @@ extension ComponentManager: PaymentComponentBuilder {
                                      context: context,
                                      configuration: .init(localizationParameters: configuration.localizationParameters))
     }
+    
+    func build(paymentMethod: StoredCashAppPayPaymentMethod) -> PaymentComponent? {
```
Comment (descorp) [inline AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift]:
```suggestion
    internal func build(paymentMethod: StoredCashAppPayPaymentMethod) -> PaymentComponent? {
```

File: Demo/Common/IntegrationExampleDropIn.swift (hunk 30 above)
Comment (erenbesel) [reply Demo/Common/IntegrationExampleDropIn.swift]: Oh nice one, we already have it. Will use `URL(string: ConfigurationConstants.returnUrl)`

File: Adyen/Core/APIClient/Environment.swift [hunk 33]
```diff
@@ -43,6 +43,12 @@ public struct Environment: AnyAPIEnvironment {
 
     /// Adyen's India live  environment.
     public static let liveIndia = Environment(baseURL: URL(string: "https://checkoutshopper-live-in.adyen.com/")!)
+    
+    /// Determines if the environment is one of Adyen's live environments.
+    @_spi(AdyenInternal)
+    public var isLive: Bool {
+        baseURL.absoluteString.hasPrefix("https://checkoutshopper-live")
```
Comment (goergisn) [inline Adyen/Core/APIClient/Environment.swift:50]: what about having a prefix property `static let liveEnvironmentPrefix: String` that we compose the live environments with - so a change on the environment urls does not accidentally break this check.

```suggestion
        baseURL.absoluteString.hasPrefix(Self.liveEnvironmentPrefix)
```

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift (hunk 29 above)
Comment (erenbesel) [reply AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift:172]: Discussed with product and this is indeed fine

File: Demo/Common/IntegrationExampleDropIn.swift (hunk 30 above)
Comment (erenbesel) [reply Demo/Common/IntegrationExampleDropIn.swift]: 6b1ab99e2e14fc8dd633aea4dac47b182b2e1794

File: Adyen/Core/Payment Methods/CashAppPaymentMethod.swift (hunk 31 above)
Comment (erenbesel) [reply Adyen/Core/Payment Methods/CashAppPaymentMethod.swift]: 6b1ab99e2e14fc8dd633aea4dac47b182b2e1794

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift (hunk 32 above)
Comment (erenbesel) [reply AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift]: 6b1ab99e2e14fc8dd633aea4dac47b182b2e1794

File: Adyen/Core/APIClient/Environment.swift (hunk 33 above)
Comment (erenbesel) [reply Adyen/Core/APIClient/Environment.swift:50]: Nice suggestion. but I'd imagine a change on live url could mean a bigger issue than this SDK :D

But a separate PR to encompass all the common part of the domain path could still be nice

**Review by goergisn — DISMISSED** [review]
Looks good 👌

File: AdyenCashAppPay/CashAppPayButton/CashAppPayButtonItem.swift [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+    
+    internal var identifier: String?
+    
+    /// The observable of the button indicator activity.
+    @AdyenObservable(false) public var showsActivityIndicator: Bool
+    
+    /// A closure that will be invoked when a button is selected.
+    internal var selectionHandler: (() -> Void)?
+    
+    internal init(selectionHandler: (() -> Void)?) {
```
Comment (goergisn) [inline AdyenCashAppPay/CashAppPayButton/CashAppPayButtonItem.swift]: Is there a reason why the selectionHandler is optional?

```suggestion
    internal init(selectionHandler: @escaping () -> Void) {
```

**Review by goergisn — DISMISSED** [review]

**Comment by mmroz** [comment]
👋 @erenbesel I think you can add `skipPackagePluginValidation` to [xcodebuild](https://github.com/Adyen/adyen-ios/blob/develop/Scripts/test-SPM-integration.sh) and that should fix CI

**Review by goergisn — APPROVED** [review]
👍

File: AdyenCashAppPay/CashAppPayComponent.swift (hunk 16 above)
Comment (erenbesel) [reply AdyenCashAppPay/CashAppPayComponent.swift]: We can't move it to init since actions are created after the button tap.

**Review by descorp — APPROVED** [review]
